multi: Add decred.

JoeGruffins opened the pull request.

JoeGruffins committed:
multi: Add decred.

Add decred coin parameters for testnet, simnet, and mainnet.

Add support for decred's unique hashing algorithm. The npm module
hash-lib is used for blake256 as it is already used in the decred
repositories.

Add a new module bs58checkBlake256 that adds support for decred encoding
and decoding. Appropriate tests are added to test all success and error
paths.

Allow decred's two byte wif prefix in networks.

JoeGruffins commented on test/fixtures/address.json:
```diff
@@ -151,6 +167,16 @@
+      {
+        "address": "DsmcYVbP1Nmag2H4AS17UTvmWXmGeA7nLD",
+        "exception": "invalid address length",
+	"coin": "DCR"
```
indent

JoeGruffins commented on test/fixtures/address.json:
```diff
@@ -151,6 +167,16 @@
+      {
+        "address": "DsmcYVbP1Nmag2H4AS17UTvmWXmGeA7nLDy",
+        "exception": "invalid checksum",
+	"coin": "DCR"
```
indent

JoeGruffins commented on test/address.js:
```diff
@@ -22,7 +23,7 @@ describe('address', function () {
       it('throws on ' + f.exception, function () {
         assert.throws(function () {
-          baddress.fromBase58Check(f.address)
+          baddress.fromBase58Check(f.address, coins[f.coin])
```
need to confirm we can't just use a string here, f.coin

JoeGruffins commented on src/networks.js:
```diff
@@ -231,5 +231,27 @@ module.exports = {
     coin: coins.KMD
+  },
+  decred: {
+    messagePrefix: 'Decred Signed Message:\n',
```
Correct? Other coins have a prefix here.

matheusd replied:
Compatible to dcrwallet, right: https://github.com/decred/dcrwallet/blob/f5b98d7c5d089fe5e0ba47bb44bc55aaad3500e8/wallet/wallet.go#L1607

JoeGruffins replied:
I see now, it's the length of the message.

JoeGruffins commented on src/networks.js:
```diff
@@ -231,5 +231,27 @@ module.exports = {
+    },
+    pubKeyHash: 0x073f,
+    scriptHash: 0x071a,
+    wif: 0,
```
wif correct?

matheusd replied:
AFAICT this is incorrect.

`wif` here seems to be the "wif version" used by bitcoinjs/wif (this is used [in ecpair](https://github.com/trezor/trezor-utxo-lib/blob/trezor/src/ecpair.js#L95) referring to [bitcoinjs/wif](https://github.com/bitcoinjs/wif/blob/78ad28eecb96f5a08a76430b24460ad48706b904/index.js#L11)).

Decred doesn't use "version" but it does have the network id sort of performing the same role. It also has a "scheme" field to specify the specific crypto algo used with that key (secp256k1, ed25519, etc) (https://github.com/decred/dcrd/blob/b60c60ffe98bcea5becc4e9e4d2a4fe6152401f6/dcrutil/wif.go#L148)

JoeGruffins replied:
thanks, found it

JoeGruffins commented on package.json:
```diff
@@ -42,7 +42,8 @@
     "bitcoin-ops": "^1.3.0",
-    "blake2b": "https://github.com/BitGo/blake2b#6268e6dd678661e0acc4359e9171b97eb1ebf8ac",
+    "blake-hash": "^1.1.0",
+    "blake2b": "git+https://github.com/BitGo/blake2b.git#6268e6dd678661e0acc4359e9171b97eb1ebf8ac",
```
should revert blake2b, this was an automatic change

JoeGruffins commented:
also need to add to the readme

JoeGruffins commented:
As you can see I am pasing coins.coin to distinguish decred from other coins. This is the biggest change and I don't know what's best. Will see what the maintainer thinks unless someone knows a better way. We could just pick decred out by looking at the leading magic bits, but that seems less straigtforward to me. It also complicates bs58check.decode because we'd have to decode once to check the magics to decide which decode to use.

matheusd commented on src/coins.js:
```diff
@@ -13,7 +13,8 @@ const coins = {
   PEERCOINTEST: 'tppc',
-  KMD: 'kmd'
+  KMD: 'kmd',
+  DCR: 'dcr'
```
If they'll accept, I'd suggest also adding testnet and simnet since it makes it easier to develop and test stuff.

vctt94 replied:
+1

JoeGruffins replied:
I don't think we need separate "coins" though. I added mainnet and testnet params, but will also add simnet too.

matheusd reviewed (commented):
Seems reasonable. See inline comments for your questions.

teknico commented on src/address.js:
```diff
@@ -1,14 +1,16 @@
 var Buffer = require('safe-buffer').Buffer
 var bech32 = require('bech32')
-var bs58check = require('bs58check')
+var bbs58check = require('./bs58check')
```
Why `bbs` instead of `bs`? What does it mean?

JoeGruffins replied:
I dunno. It is a convention already in place to put a "b" when using local modules. base already starts with a b so, bbs...

vctt94 reviewed (approved):
LGTM

JoeGruffins committed:
Add decred transactions.

 - Add from/toBuffer functionality.

 - Add transaction copying.

 - Add hash for sig for all hash types. This includes separate functions
   to calculate decred prefix and witness lengths. Add tests for all
   hash types.

 - Add fromTransaction to transaction builder.

 - Add adding decred inputs, outputs, and witnesses. Inputs and
   witnesses come with functions that are separated from other coins.

 - Add signing. Signing is done with a separate function than other
   coins. Add a separate signing test.

JoeGruffins committed:
hdnode: Use blake256 for decred.

JoeGruffins committed:
Make networks more distinct.

JoeGruffins committed:
Fix node v6.

JoeGruffins committed:
Create ops.js and add staking scripts.

JoeGruffins committed:
Revert transaction builder changes.

JoeGruffins committed:
Fix comment.
